File: backend/app/core/branch_tree.py
"""Ported from packages/core/src/branchTree.ts"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional

from app.domain.types import ScriptBlock, VnProject


@dataclass
class BranchNode:
    id: str
    kind: str  # "label" | "menu" | "choice" | "jump" | "end"
    title: str
    children: List["BranchNode"] = field(default_factory=list)


def build_branch_tree(project: VnProject, chapterId: Optional[str] = None) -> List[BranchNode]:
    """Build a lightweight branch tree from menu/jump/label structure."""
    chapters = (
        [c for c in project.chapters if c.id == chapterId] if chapterId else project.chapters
    )

    roots: List[BranchNode] = []
    for ch in chapters:
        chapter_root = BranchNode(
            id=f"ch-{ch.id}",
            kind="label",
            title=ch.title,
            children=_walk_blocks(ch.blocks, ch.id),
        )
        roots.append(chapter_root)
    return roots


def _walk_blocks(blocks: List[ScriptBlock], prefix: str) -> List[BranchNode]:
    nodes: List[BranchNode] = []
    for i, b in enumerate(blocks):
        btype = b.get("type")
        if btype == "label":
            nodes.append(
                BranchNode(
                    id=f"{prefix}-label-{b['name']}-{i}",
                    kind="label",
                    title=f"label {b['name']}",
                    children=[],
                )
            )
        elif btype == "menu":
            choices = b.get("choices", []) or []
            children: List[BranchNode] = []
            for ci, c in enumerate(choices):
                jump = c.get("jump")
                children.append(
                    BranchNode(
                        id=f"{prefix}-choice-{b.get('id')}-{ci}",
                        kind="choice",
                        title=c.get("text", ""),
                        children=(
                            [
                                BranchNode(
                                    id=f"{prefix}-jump-{jump}-{ci}",
                                    kind="jump",
                                    title=f"→ {jump}",
                                    children=[],
                                )
                            ]
                            if jump
                            else []
                        ),
                    )
                )
            prompt = b.get("prompt")
            menu_node = BranchNode(
                id=f"{prefix}-menu-{b.get('id')}-{i}",
                kind="menu",
                title=f"选项：{prompt}" if prompt else f"menu {b.get('id')}",
                children=children,
            )
            nodes.append(menu_node)
        elif btype == "jump":
            nodes.append(
                BranchNode(
                    id=f"{prefix}-jump-{b['target']}-{i}",
                    kind="jump",
                    title=f"jump {b['target']}",
                    children=[],
                )
            )
        elif btype == "return":
            nodes.append(
                BranchNode(id=f"{prefix}-end-{i}", kind="end", title="return", children=[])
            )
    return nodes

File: frontend/src/lib/branchTree.ts
import type { ScriptBlock, VnProject } from "../types/vn";

export interface BranchNode {
  id: string;
  kind: "label" | "menu" | "choice" | "jump" | "end";
  title: string;
  children: BranchNode[];
}

/** Build a lightweight branch tree from menu/jump/label structure. */
export function buildBranchTree(project: VnProject, chapterId?: string): BranchNode[] {
  const chapters = chapterId
    ? project.chapters.filter((c) => c.id === chapterId)
    : project.chapters;

  const roots: BranchNode[] = [];

  for (const ch of chapters) {
    const chapterRoot: BranchNode = {
      id: `ch-${ch.id}`,
      kind: "label",
      title: ch.title,
      children: walkBlocks(ch.blocks, ch.id),
    };
    roots.push(chapterRoot);
  }
  return roots;
}

function walkBlocks(blocks: ScriptBlock[], prefix: string): BranchNode[] {
  const nodes: BranchNode[] = [];
  let i = 0;
  while (i < blocks.length) {
    const b = blocks[i];
    if (b.type === "label") {
      nodes.push({
        id: `${prefix}-label-${b.name}-${i}`,
        kind: "label",
        title: `label ${b.name}`,
        children: [],
      });
    } else if (b.type === "menu") {
      const menuNode: BranchNode = {
        id: `${prefix}-menu-${b.id}-${i}`,
        kind: "menu",
        title: b.prompt ? `选项：${b.prompt}` : `menu ${b.id}`,
        children: b.choices.map((c, ci) => ({
          id: `${prefix}-choice-${b.id}-${ci}`,
          kind: "choice" as const,
          title: c.text,
          children: c.jump
            ? [
                {
                  id: `${prefix}-jump-${c.jump}-${ci}`,
                  kind: "jump" as const,
                  title: `→ ${c.jump}`,
                  children: [],
                },
              ]
            : [],
        })),
      };
      nodes.push(menuNode);
    } else if (b.type === "jump") {
      nodes.push({
        id: `${prefix}-jump-${b.target}-${i}`,
        kind: "jump",
        title: `jump ${b.target}`,
        children: [],
      });
    } else if (b.type === "return") {
      nodes.push({
        id: `${prefix}-end-${i}`,
        kind: "end",
        title: "return",
        children: [],
      });
    }
    i += 1;
  }
  return nodes;
}
